src/packer.pkr.hcl

```
@@ -0,0 +1,131 @@
+variable "ami_regions" {
+  default     = []
+  description = "The list of AWS regions to copy the AMI to once it has been created. Example: [\"us-east-1\"]"
+  type        = list(string)
+}
+
+variable "build_region" {
+  default     = "us-east-1"
+  description = "The region in which to retrieve the base AMI from and build the new AMI."
+  type        = string
+}
+
+variable "build_region_kms" {
+  default     = "alias/cool-amis"
+  description = "The ID or ARN of the KMS key to use for AMI encryption."
+  type        = string
+}
+
+variable "is_prerelease" {
+  default     = "false"
+  description = "The pre-release status to use for the tags applied to the created AMI."
+  type        = string
```

Shouldn't this be a boolean?
```suggestion
  type        = bool
```

---

We usually do `map(string)` for AWS tags, so I thought it should be left as a string since it is only used for tagging. If desired I can look at converting it to a `bool`, but that will require some massaging of the workflows to avoid empty input values.

---

If it's a big hassle, I'm fine leaving it as a string, but in that case, we should mention in the variable description that valid values are "true" and "false".

---

I vote for changing it to a `bool` unless it's a really heavy lift.  As an input variable it is user-facing and therefore should be a simple as possible.  A `bool` is definitely simpler (and less error prone) than a `string` that takes on the two values `true` and `false`.

If it is a really heavy lift and we need to leave it as a `string` then please add a comment explaining why that is the case; otherwise in a month I will have no memory of this interaction and I will suggest changing it to a `bool` again.

---

The main issue is in the workflows the value of `github.event.release.prerelease` is empty if it isn't a pre-release/release. We need to ensure it's a valid string version of a boolean value before it can be passed in to the command line. The quick lift is to remove all references to `github.event.release` in the `build.yml` workflow because it is both not creating an AMI and not designed to process a pre-release/release, so you end up with

```console
  packer build -timestamp-ui \
    -var is_prerelease= \
    -var release_tag= \
    -var release_url= \
    -var skip_create_ami=true \
    src/packer.pkr.hcl
```

when the workflow runs (per the [logs](https://github.com/cisagov/skeleton-packer/runs/5187675424?check_suite_focus=true#step:13:2-7)). Instead alternate values could be inserted/defaults allowed to be used.

---

OK, then leave it as a string but add the validation rules to ensure that only `"true"` and `"false"` are accepted for the variable.  Also make sure that the variable description tells the user what the accepted values are, including issues of case (`"true"` vs `"True"`).

---

What @jsf9k said works for me.

---

Adding validation or switching to a `bool` will both require adjusting the workflow because neither allows an empty value to be passed. I would rather switch to a `bool` and remove the references to the `github.event.release` object from the `build.yml` workflow. No image is created so setting tag values is not necessary (the defaults are fine). A bool friendly value should be passed in when the `prerelease.yml` and/or `release.yml` workflow run because they should have a `github.event.release` object.

---

Why does adding validation require adjusting the workflow?  Can't you just use [this](https://www.packer.io/docs/templates/hcl_templates/variables#custom-validation-rules)?

---

Never mind, it's the empty values that are the problem.

---

> I would rather switch to a `bool` and remove the references to the `github.event.release` object from the `build.yml` workflow. No image is created so setting tag values is not necessary (the defaults are fine). A bool friendly value should be passed in when the `prerelease.yml` and/or `release.yml` workflow run because they should have a `github.event.release` object.

I like this solution.  Can you either do this or create an issue as a reminder to do it later?

---

Please see [d159adc](https://github.com/cisagov/skeleton-packer/pull/98/commits/d159adcfe324de818da7a0e0f440807072a2b95b). I created a new pre-release and confirmed that an AMI was successfully created and that the tags are correctly set on the AWS console.

---

That looks much better! 👍

---

@jsf9k - Let us know if https://github.com/cisagov/skeleton-packer/pull/98/commits/d159adcfe324de818da7a0e0f440807072a2b95b works for you or if you have any concerns.  Once I hear from you, I'll resolve this conversation (or feel free to just resolve it yourself).